res/controllers/midi-components-0.0.js

```diff
@@ -27,14 +27,63 @@
  */
 
 (function(global) {
+    /**
+     * @typedef {[number, number] | [number, [number, number]] | [[number, number], number] | [[number, number], [number, number]]} MidiBytesVariations
+     */
+
+    /**
+     * @typedef {object} MidiBytesAsObject
+     * @property {MidiBytesVariations} [in] MIDI signals from device to listen
+     * @property {MidiBytesVariations} [out] MIDI signals to transmit to the device
+     */
+
+    /**
+     * @typedef {object} ComponentOptions
+     * @property {MidiBytesAsObject | [number, number]} [midi] MIDI events
+     */
+
     const NO_TIMER = 0;
+    /**
+     * @param {ComponentOptions | [number, number]} options Component configuration
+     * @class
+     */
     const Component = function(options) {
-        if (Array.isArray(options) && typeof options[0] === "number") {
-            this.midi = options;
+        if (Array.isArray(options)) {
+            const [midiStatus, midiNo] = options;
+            options = {midi: {out: [midiStatus, midiNo]}};
+        } else if (Array.isArray(options.midi)) {
+            options.midi = {out: options.midi};
         } else {
-            Object.assign(this, options);
+            options.midi = {out: undefined, in: undefined};
         }
 
+        const normalizeMidi = (midiBytesVar) => {
+            if (!Array.isArray(midiBytesVar)) {
+                return [];
+            }
+
+            const [first, second] = midiBytesVar;
+            const firstIsArray = Array.isArray(first);
+            const secondIsArray = Array.isArray(second);
+
+            if (firstIsArray && secondIsArray) { // long form
+                return midiBytesVar;
+            } else if (!firstIsArray && !secondIsArray) {
+                return [midiBytesVar];
+            } else if (Array.isArray(first)) { // [[number, number], number] form
+                const [status1, status2] = first;
+                return [[status1, second], [status2, second]];
+            } else { // [number, [number, number]] form
+                const [midino1, midino2] = second;
+                return [[first, midino1], [first, midino2]];
+            }
+        };
```

Comment:
lets undo this for now. Its IMO unnecessary complexity compared to the fuller blown solution. We should focus on other issues first.